### PR TITLE
Active contour callback

### PR DESCRIPTION
**Pull request by noahstier**
## Description

Add an option for a callback function in active_contour, in skimage/segmentation/active_contour_model.py
## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [n/a] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [n/a] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests
## References

implements enhancement #2092 
adds documentation #2093

### CODE REVIEW

**Comment by noahstier** [comment]
`if not new_scipy: raise NotImplementedError` at https://github.com/scikit-image/scikit-image/blob/master/skimage/segmentation/active_contour_model.py#L92 makes the `else` of `if new_scipy: ... else:` unreachable at https://github.com/scikit-image/scikit-image/blob/master/skimage/segmentation/active_contour_model.py#L131 and https://github.com/scikit-image/scikit-image/blob/master/skimage/segmentation/active_contour_model.py#L189 

Using `@skipif(not new_scipy)` for the tests seems to dodge this issue. If someone is running these tests using an unsupported version of scipy, shouldn't the tests fail?

File: skimage/segmentation/active_contour_model.py [hunk 1]
```diff
@@ -9,20 +9,26 @@
...
+                   callback=None):
     """Active contour model.
 
     Active contours by fitting snakes to features of images. Supports single
     and multichannel 2D images. Snakes can be periodic (for segmentation) or
     have fixed and/or free ends.
 
+    The input snake is a set of points that will each be repositioned such that
+    the new snake minimizes energy as described in [1]_. No points are added
+    or removed.
```
Comment (emmanuelle) [inline skimage/segmentation/active_contour_model.py]: How about 
No points are added or removed --> The number of points is constant, so make sure that the initial snake has enough points to capture the details of the final contour.

File: skimage/segmentation/active_contour_model.py [hunk 2]
```diff
@@ -9,20 +9,26 @@
...
+
     Parameters
     ----------
     image : (N, M) or (N, M, 3) ndarray
         Input image.
     snake : (N, 2) ndarray
         Initialisation coordinates of snake. For periodic snakes, it should
         not include duplicate endpoints.
+        N is not necessarily the same N as the input image dimension.
```
Comment (emmanuelle) [inline skimage/segmentation/active_contour_model.py]: use a different letter then, it shouldn't be ``N``.

File: skimage/segmentation/active_contour_model.py [hunk 3]
```diff
@@ -48,6 +54,16 @@ def active_contour(image, snake, alpha=0.01, beta=0.1,
         Maximum iterations to optimize snake shape.
     convergence: float, optional
         Convergence criteria.
+    callback: function, optional
+        Called every 11 iterations with three args.
```
Comment (emmanuelle) [inline skimage/segmentation/active_contour_model.py]: 11 is a strange number. (Why not 10?) In fact, it is called every time convergence is tested as well. Should we say this instead?

**Comment by emmanuelle** [comment]
Thanks for the addition and the other fixes. I left a couple of minor comments.

**Comment by noahstier** [comment]
@emmanuelle I agree with your suggestions, thanks. 

As for 
> 11 is a strange number
the idea is that the snake moves with a fixed step size, and convergence is checked by comparing a to the previous 10 snakes, leading to an 11-iteration cycle. Fitting the callback into that cycle seemed like an efficient way to do it.

After some research I found that a fixed step size has been known not to work very well, since like 20 years ago. (Fua & Leclerc: Model Driven Edge Detection, p. 48). Instead a better way to do it is to check convergence every iteration, using a variable step size.

Another guy recommends updating step size approximately every 10 iterations: Liew, Visual Speech Recognition: Lip Segmentation and Mapping: Lip Segmentation and Mapping p.193 .

This is something I'd be interested in implementing, but before I go for it, @juliusbierk maybe you can shed some light on this?

**Comment by stefanv** [comment]
Kudos on that piece of research, @noahstier.

File: skimage/segmentation/active_contour_model.py (hunk 1 above)
Comment (stefanv) [inline skimage/segmentation/active_contour_model.py]: :+1:

File: skimage/segmentation/active_contour_model.py (hunk 3 above)
Comment (stefanv) [inline skimage/segmentation/active_contour_model.py]: Why not just call it each iteration?  The user can decide to skip over 10, if they choose to.

File: skimage/segmentation/active_contour_model.py [hunk 4]
```diff
@@ -48,6 +54,16 @@ def active_contour(image, snake, alpha=0.01, beta=0.1,
         Maximum iterations to optimize snake shape.
     convergence: float, optional
         Convergence criteria.
+    callback: function, optional
```
Comment (stefanv) [inline skimage/segmentation/active_contour_model.py]:
```
callback : f(snake, n, conv)
```

Then describe those in the docstring.

File: skimage/segmentation/active_contour_model.py [hunk 5]
```diff
@@ -48,6 +54,16 @@ def active_contour(image, snake, alpha=0.01, beta=0.1,
         Maximum iterations to optimize snake shape.
     convergence: float, optional
         Convergence criteria.
+    callback: function, optional
+        Called every 11 iterations with three args.
+        1st arg: (N, 2) ndarray -- the current snake.
+        2nd arg: int -- the total number of iterations completed.
+        3rd arg: float -- convergence criterion. This value is computed by
```
Comment (stefanv) [inline skimage/segmentation/active_contour_model.py]: We should be quite precise in describing how the convergence is computed.  Let's wait for your other research to figure out what exactly it should be.

File: skimage/segmentation/active_contour_model.py [hunk 6]
```diff
@@ -84,6 +100,17 @@ def active_contour(image, snake, alpha=0.01, beta=0.1,
     >>> int(np.mean(dist)) #doctest: +SKIP
     25
 
+    Watch the snake progress:
+
+    >>> import matplotlib.pyplot as plt
```
Comment (stefanv) [inline skimage/segmentation/active_contour_model.py]: If I'm not mistaken, plt should already be available in the doctest namespace.

Comment (stefanv) [inline skimage/segmentation/active_contour_model.py]: Actually, disregard that--I think it's good to be explicit here.

**Comment by emmanuelle** [comment]
@noahstier I would suggest that we converge on this PR, so that we can merge the callback part. The issue with fixed or adaptive step can be fixed in another PR later.

**Comment by noahstier** [comment]
I believe this is finished.

Not sure what else to do about the failing tests though. `python -c "import skimage; skimage.test(doctest=True)"` and everything passes locally. However travis is complaining about unexpected indent in the docstring in `active_contour_model.py`, and appveyor can't find fblas. Maybe someone can take a look?

**Comment by noahstier** [comment]
closing in favor of https://github.com/scikit-image/scikit-image/pull/2586

**Pull request closed without merging**